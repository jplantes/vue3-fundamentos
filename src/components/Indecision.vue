<template>
  <!-- v-baind: -> : -->
  <img v-if="img" :src="img" alt="bg">

  <div class="bg-dark"></div>

  <div class="indecision-container">

    <input 
      type="text" 
      placeholder="Hazme una pregunta"
      v-model="pregunta"
    >
    <p>Recuerda termina con un signo de interrogaci√≥n (?)</p>
    
    <div v-if="isValidCuestion" >
      <h2>{{ pregunta }}</h2>
      <h1>{{ respuesta }}</h1>
    </div>

  </div>

</template>

<script>
  export default {
  
    data() {
      return {
        pregunta: null,
        respuesta: null,
        img: null,
        isValidCuestion: false
      }
    },

    methods: {
      async getPregunta() {

        try {
          this.respuesta = 'Pensando...'
  
          const { answer, image } = await fetch( 'https://yesno.wtf/api' ).then( r => r.json() )
  
          this.respuesta = ( answer === 'yes' ) ? 'Si' : 'No' 
          this.img = image
        } catch (error) {
          this.respuesta = 'No se pudo cargar del API'
          
        }
      }
    },

    watch: {
      // El observable se tiene que llamar igual que la propiedad que queremos observar
      pregunta( value, oldValue ) {

        this.isValidCuestion = false

        console.log({ value });

        if ( !value.includes( '?' ) ) return

        this.isValidCuestion = true

        //TODO: hacer peticion http
        this.getPregunta()

      }
    }


  }
</script>

<style lang="css" scoped>
  img, .bg-dark {
      height: 100vh;
      left: 0px;
      max-height: 100%;
      max-width: 100%;
      position: fixed;
      top: 0px;
      width: 100vw;
  }

  .bg-dark {
      background-color: rgba(0, 0, 0, 0.4);
  }

  .indecision-container {
      position: relative;
      z-index: 99;
  }
  
  input {
      width: 250px;
      padding: 10px 15px;
      border-radius: 5px;
      border: none;
  }
  input:focus {
      outline: none;
  }

  p {
      color: white;
      font-size: 20px;
      margin-top: 0px;
  }

  h1, h2 {
      color: white;
  }
  
  h2 {
      margin-top: 150px;
  }
</style>