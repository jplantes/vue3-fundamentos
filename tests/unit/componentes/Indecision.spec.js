import { shallowMount } from "@vue/test-utils";

import Indecision from '@/components/Indecision'

describe( 'Pruebas sobre el componente <Indecision />', () => {

  let wrapper;

  // Para evaluar el console.log
  let clgSpy;

  //   Implementamos un "mock" sobre fetchAPI ya que en nodeJS no existe
  // y para solucionar el error del .then le decimos que devuelve una 
  // Promesa resuelta
  // Dentro de la resolucion retornamos el json
  global.fetch = jest.fn( () => Promise.resolve({
    json: () => Promise.resolve({
      answer: 'yes',
      forced: false,
      image: 'https://yesno.wtf/assets/yes/2.gif'
    })
  }));

  beforeEach( () => {
    wrapper = shallowMount( Indecision );
    clgSpy = jest.spyOn( console, 'log' );

    // Limpiar todos los mocks
    jest.clearAllMocks();
  });

  test( 'Debe coincidir con el snapshot', () => {

    expect( wrapper.html() ).toMatchSnapshot();
    
  });

  test( 'Escribir en el input no debe disparar nada', async () => {

    // El .vm es la instancia de VUE para poder acceder a las funciones
    const getPreguntasSpy = jest.spyOn( wrapper.vm, 'getPregunta' );

    const inputHtml = wrapper.find( 'input' );
    await inputHtml.setValue( 'Hola mundo' );

    expect( clgSpy ).toHaveBeenCalled();
    expect( getPreguntasSpy ).not.toHaveBeenCalled();
    
  });

  test( 'Escribir en el input (?) debe disparar la peticiÃ³n', async () => {
    // El .vm es la instancia de VUE para poder acceder a las funciones
    const getPreguntasSpy = jest.spyOn( wrapper.vm, 'getPregunta' );

    const inputHtml = wrapper.find( 'input' );
    await inputHtml.setValue( 'Hola mundo?' );

    expect( clgSpy ).toHaveBeenCalled();
    expect( getPreguntasSpy ).toHaveBeenCalled();
  });

  test( 'Pruebas en getPregunta', async () => {

    await wrapper.vm.getPregunta();

    const imgHtml = wrapper.find( 'img' );
    expect( imgHtml.exists() ).toBeTruthy();


    expect( wrapper.vm.img ).toBe( 'https://yesno.wtf/assets/yes/2.gif' );
    expect( wrapper.vm.respuesta ).toBe( 'Si' );

  });

  test( 'Pruebas en getPregunta - fallo en el API', async () => {
    
    fetch.mockImplementationOnce( () => Promise.reject( 'API is down' ));
    
    await wrapper.vm.getPregunta();

    const imgHtml = wrapper.find( 'img' );
    expect( imgHtml.exists() ).toBeFalsy();
    expect( wrapper.vm.respuesta ).toBe( 'No se pudo cargar del API' );
  });

});