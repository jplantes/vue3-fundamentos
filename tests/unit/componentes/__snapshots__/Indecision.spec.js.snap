// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre el componente <Indecision /> Debe coincidir con el snapshot 1`] = `
<!-- v-baind: -> : -->
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda termina con un signo de interrogaci√≥n (?)</p>
  <!--v-if-->
</div>
`;
